inject.sh

```diff
@@ -1,85 +1,90 @@
-         -ex "set \$self = \$dlopen(\"$libpath\", 6)"       \
+         -ex "set \$self = \$dlopen(\"$tmp_libpath\", 6)"   \
          -ex "call \$dlclose(\$self)"                       \
          -ex "call \$dlclose(\$self)"                       \
                                                             \
          -ex "call \$dlerror()"                             \
          -ex "detach"                                       \
          -ex "quit"
-
-    exit 0
-elif [ "$1" == "debug" ]; then
+elif [[ "$1" == "debug" ]]; then
    echo "Launching in debug mode."
-   sudo gdb -n -q                                          \
+   gdb -n -q                                               \
         -ex "attach $pid"                                  \
         -ex "set \$dlopen = (void* (*)(char*, int))dlopen" \
         -ex "set \$dlclose = (int (*)(void*))dlclose"      \
         -ex "set \$dlerror =  (char* (*)(void))dlerror"    \
-        -ex "call \$dlopen(\"$libpath\", 2)"               \
+        -ex "call \$dlopen(\"$tmp_libpath\", 2)"           \
         -ex "call \$dlerror()"                             \
         -ex "continue" # Comment this line for manual debug
 
-    exit 0
-fi
-
-if grep -q "$libpath" "/proc/$pid/maps"; then
-    echo -e "Enoch already loaded. Reloading...\n";
-
-    # 0x2 -> RTLD_NOW
-    # 0x6 -> RTLD_LAZY | RTLD_NOLOAD
-    # For more info on the 3 mid lines, see self_unload() in main.c
-    sudo gdb -n -q -batch                                   \
-         -ex "attach $pid"                                  \
-         -ex "set \$dlopen = (void* (*)(char*, int))dlopen" \
-         -ex "set \$dlclose = (int (*)(void*))dlclose"      \
-         -ex "set \$dlerror =  (char* (*)(void))dlerror"    \
-                                                            \
-         -ex "set \$self = \$dlopen(\"$libpath\", 6)"       \
-         -ex "call \$dlclose(\$self)"                       \
-         -ex "call \$dlclose(\$self)"                       \
-                                                            \
-         -ex "call \$dlopen(\"$libpath\", 2)"               \
-         -ex "call \$dlerror()"                             \
-         -ex "detach"                                       \
-         -ex "quit"
+    set +x
+    echo -e "\nDone. Unload cheat using:\n$0 unload $tmp_libpath"
+    rm "$tmp_libpath"
 else
-    sudo gdb -n -q -batch                                   \
-         -ex "attach $pid"                                  \
-         -ex "set \$dlopen = (void* (*)(char*, int))dlopen" \
-         -ex "set \$dlclose = (int (*)(void*))dlclose"      \
-         -ex "set \$dlerror =  (char* (*)(void))dlerror"    \
-         -ex "call \$dlopen(\"$libpath\", 2)"               \
-         -ex "call \$dlerror()"                             \
-         -ex "detach"                                       \
-         -ex "quit"
+    # Im still thinking of a way to reload cheat while using temp library
```

Comment:
Maybe we could save the name of the temp library in use in another temporary file (e.g. `/tmp/enoch-lib-name.txt`) and unload by checking if the file exists, and using its contents. It's hacky, but it's an idea.

When we are done (e.g. after unloading), we could delete that file. The contents of the file can be overwritten when loading the cheat by just using `echo "..." > /tmp/...`

Reply:
Hmm, what about using /tmp/hl2_linux_$pid_enoch.txt ? so it will work on multiple tf2 instances (in theory) too

Reply:
The only problem would be leaving the files there after closing the game.

Imagine the following:
1. Open instance A with PID 123.
2. Inject. File `123_enoch.txt` gets generated, containing name of random lib A.
3. Open instance B with PID 456.
4. Inject. File `456_enoch.txt` gets generated, containing name of random lib B.
5. Unload from instance A, since it has PID 123, removes `123_enoch.txt` after reading it.
6. Close instance B without unloading.
7. Next instances won't have PID 456, and the file can't be deleted.

It's not a huge problem since they are in the `/tmp` folder, but still, even more hacky.

Reply:
Thats ok, worst thing that can happen is if VAC will check if those hl2_linux_${pid}_enoch.txt files exist!

Reply:
That would be an honor. :)